File: frontend/src/api/statistics.ts
import type { TimePeriod } from '@/types/statistics.types'
import { api } from './client'

export const getIncomeExpenseOverview = async (
  start: Date,
  end: Date,
  period: TimePeriod,
  user_timezone: string,
) => {
  const { data } = await api.get('/statistics/income-expense', {
    params: {
      start,
      end,
      period,
      user_timezone,
    },
  })
  return data
}

export const getSpendingByCategory = async (start: Date, end: Date) => {
  const { data } = await api.get('/statistics/spending-by-category', {
    params: {
      start,
      end,
    },
  })
  return data
}

export const getFinancialSummary = async (user_timezone: string) => {
  const { data } = await api.get('/statistics/financial-summary', {
    params: {
      user_timezone,
    },
  })
  return data
}

File: backend/app/routers/statistics.py
from datetime import datetime

from fastapi import APIRouter

from app.db.database import SessionDep
from app.services import statistics
from app.schemas.statistics import (
    FinancialSummary,
    IncomeExpenseOverview,
    SpendingByCategory,
)
from app.dependencies import CurrentUser

router = APIRouter(prefix="/statistics")


@router.get("/income-expense", response_model=list[IncomeExpenseOverview])
def get_income_expense_overview(
    session: SessionDep,
    user: CurrentUser,
    start: datetime,
    end: datetime,
    period: str,
    user_timezone: str,
):
    stats = statistics.get_income_expense_overview(
        session, start, end, period, user_timezone
    )
    return stats


@router.get("/spending-by-category", response_model=list[SpendingByCategory])
def get_spending_by_category(
    session: SessionDep,
    user: CurrentUser,
    start: datetime,
    end: datetime,
):
    stats = statistics.get_spending_by_category(session, start, end)
    return stats


@router.get("/financial-summary", response_model=FinancialSummary)
def get_financial_summary(session: SessionDep, user: CurrentUser, user_timezone: str):
    stats = statistics.get_financial_summary(session, user_timezone)
    return stats
